### ISSUE TITLE
[Spell] Bless

### ISSUE THREAD

**Issue by FelipeRenault** [ZerdBG3/RAW#89]
Test if it's possible to make Bless have a clear target selection, instead of selecting the closest to the center

**Comment by Arkenny** [comment]
Bane and Aid come to mind as well, fix Aid amount of targets as well.

**Comment by DubhHerder** [comment]
I got a working version of 3 target of bless. It using magic missile as a template
 
https://github.com/DubhHerder/DnD-Rebalancing/blob/changingBless/DnDRebalancing/Public/DnDRebalancing/Stats/Generated/Data/Spell_Projectile.txt

not sure how you would like it to be used but it seems to resolve the issue

**Comment by FelipeRenault** [comment]
I'm scared of it being a projectile, though. I'll test if the property `AmountOfTargets` can be used on Target spells first

**Comment by DubhHerder** [comment]
I attempted the same and wasn't able to target multiple allies when I combined `amountoftargets` with a Target version of the spell but i was changing multiple parameters while testing so it might be possible for you to do. For that combination I could only target a single ally.

Can you expand on why changing it to projectile is not a great solution?

**Comment by FelipeRenault** [comment]
Projectiles can possibly be blocked by terrain and other targets in front of the selected target. It can be a possible solution if we find a template which doesn't actually shoot any projectile and just automatically provides the status to the targets

**Comment by DubhHerder** [comment]
Yeah that's a good reason to avoid this implementation, I hadn't realised the direct line of sight limitation was integrated into projectiles

**Comment by FelipeRenault** [comment]
So, I tested a few alternatives. `AmountOfTargets` can't be used on `Target` spells, so I also started testing with `Projectile`.
If you set `ProjectileCount` to 0 or remove the `Template`, the spell goes off and doesn't shoot anything, causing the buff to not go through.
The next best option is to use an invisible projectile `Template`. I've found this one `b64b516c-1afd-4f8d-b624-4c9caf06f1c2`, which is used for some trap activations (most notably, the rolling boulder damage). However, even with an invisible `Template`, the projectile path and travel time is still considered, meaning there's potential for a terrain to block the buff, or even an enemy standing in front of your ally stealing the buff - we can prevent the buff being applied to an enemy, but they might block the projectile regardless, at least denying the buff going through

**Comment by FelipeRenault** [comment]
I'll register here my best attempt into making Bless a projectile
```
new entry "Target_Bless"
type "SpellData"
data "SpellType" "Projectile"
data "Level" "1"
data "SpellSchool" "Enchantment"
data "SpellProperties" "ApplyStatus(BLESS, 100, 10)"
data "TargetRadius" "9"
data "AmountOfTargets" "3"
data "TargetConditions" "not Item() and not Dead() and not Enemy()"
data "ProjectileCount" "1"
data "Template" "b64b516c-1afd-4f8d-b624-4c9caf06f1c2"
data "Icon" "Spell_Enchantment_Bless"
data "DisplayName" "Target_Bless_DisplayName"
data "Description" "Target_Bless_Description"
data "TooltipStatusApply" "ApplyStatus(BLESS, 100, 10)"
data "PreviewCursor" "Cast"
data "CastTextEvent" "Cast"
data "CastSound" "Spell_Cast_Buff_Bless_L1to3"
data "CycleConditions" "Ally() and not Dead()"
data "UseCosts" "ActionPoint:1;SpellSlot:1:1:1"
data "SpellAnimationArcaneMagic" "414bbf02-2918-4f01-83fb-1ddc7a588d88(SPL_Arcane_Healing_01_Prepare),,;,,;09ae2f11-f5b4-42f5-ae16-687a5b57d500(SPL_Somatic_Aoe_01_Cast),,;10caea0e-c949-4d91-8ab7-3b50019dd054(SPL_Somatic_Aoe_01_Recover),,;,,;,,"
data "SpellAnimationDivineMagic" "414bbf02-2918-4f01-83fb-1ddc7a588d88(SPL_Arcane_Healing_01_Prepare),,;,,;09ae2f11-f5b4-42f5-ae16-687a5b57d500(SPL_Somatic_Aoe_01_Cast),,;10caea0e-c949-4d91-8ab7-3b50019dd054(SPL_Somatic_Aoe_01_Recover),,;,,;,,"
data "SpellAnimationNoneMagic" "414bbf02-2918-4f01-83fb-1ddc7a588d88(SPL_Arcane_Healing_01_Prepare),,;,,;09ae2f11-f5b4-42f5-ae16-687a5b57d500(SPL_Somatic_Aoe_01_Cast),,;10caea0e-c949-4d91-8ab7-3b50019dd054(SPL_Somatic_Aoe_01_Recover),,;,,;,,"
data "VerbalIntent" "Buff"
data "SpellStyleGroup" "Class"
data "SpellFlags" "HasVerbalComponent;HasSomaticComponent;IsConcentration;IsSpell"
data "RechargeValues" "6"
data "HitAnimationType" "MagicalNonDamage"
data "MemoryCost" "1"
data "PrepareEffect" "60ec0a2d-1b72-4fb2-9eaf-571638aca25d"
data "CastEffect" "6d072523-11f5-4b1b-a2d8-24ade34a315d"
data "TargetEffect" "4df0bae6-fccf-48a0-804c-3edc40264d07"
data "PositionEffect" "2832c7dd-14bd-4c51-9d2a-599c2c19005f"
```

**Comment by DubhHerder** [comment]
It looks like the difference between the two templates is how the projectile path is specified via `PathMaxArcDist`, `PathRadius`, `ProjectilePath` and travel time via `Speed`.

It might be possible to tweak those values in a new root template to optimise the projectile paths and travel time.

I guess such curving paths are an attempt to remove direct line of sight limitations for magic missile and the invisible template you found

**Comment by DubhHerder** [comment]
An alternative solution (that is probably worse) could be to cast bless on one target and have it add extra spells that dont cost a spell slot to cast. This is the approach taken by produce flame and twinned spell in the sorcerer mod on nexus. It would be technically more accurate but the user experience would likely be poorer, especially with how the hotbar works.

**Comment by FelipeRenault** [comment]
Yeah, that might work. I can even use the solution I've made to Produce Flame, so the bless caster always has the "free bless" cast, but can't use it until the first main cast is used (costing a spell slot) and then provides a status to use the free one X times. I'm not sure if that'd work given Bless already is a spell container when you get second level spell slots.
But regardless, I'd avoid doing that, because casting Bless individually 4 times (when lvl 2) seems very tedious and slow

**Comment by DubhHerder** [comment]
Agreed.

I haven't been able to find a list of what attributes are available for a RootTemplate such as `PathMaxArcDist`,  `PathRadius`, `ProjectilePath` etc. I was hoping that looking through the possible attributes might give additional options for modifying a template for Bless. Do you know if such a list can be found anywhere?

**Comment by FelipeRenault** [comment]
I've never found a compiled list for that, no

**Comment by DubhHerder** [comment]
That's a shame. I guess then the choice is to go with one of the suggested solutions or put the issue on hold and stick with base version until we see how they handle twinned spell metamagic (or another similar ability). They might add more functionality or they might take the current upcasting approach to switch from a single target to an area target .

**Comment by FelipeRenault** [comment]
I was hoping Twinned Spell from Sorcerer could provide an useful function to use for multiple target selection with Target spells, but here's how they did it:
```
data "Boosts" "UnlockSpellVariant(TwinnedProjectileSpellCheck(),ModifyNumberOfTargets(AdditiveBase,1,false),ModifyUseCosts(Add,SorceryPoint,SpellPowerLevel,0));UnlockSpellVariant(TwinnedTargetSpellCheck(),ModifyAreaRadius(AdditiveBase,6),ModifyMaximumTargets(Override,2),ModifyUseCosts(Add,SorceryPoint,SpellPowerLevel,0));UnlockSpellVariant(TwinnedTargetTouchSpellCheck(),ModifyAreaRadius(AdditiveBase,2),ModifyMaximumTargets(Override,2),ModifyUseCosts(Add,SorceryPoint,SpellPowerLevel,0))"
```
I didn't test it in-game, but from what's written there it looks like for the `TwinnedTargetSpellCheck()` (which checks if the spell is a Target one, doesn't have an area of effect and has a range bigger than touch) does this:
```
UnlockSpellVariant(TwinnedTargetSpellCheck(),ModifyAreaRadius(AdditiveBase,6),ModifyMaximumTargets(Override,2),ModifyUseCosts(Add,SorceryPoint,SpellPowerLevel,0))
```
translating to: modify the Area Radius to a flat 6 and makes the target amount 2, while also adding the sorcery point cost. This means that even twinned spell can't make a target spell have a selector, instead using the current area auto-selection that Bless has.
It's very unfortunate :/

**Comment by FelipeRenault** [comment]
Larian actually made Target spells have a `AmountOfTargets` and already used that for Bless.
FINALLY!

**Comment by Arkenny** [comment]
It like i am not crazy when i made the test lol ;P.

**Comment by FelipeRenault** [comment]
Well, maybe they got mad that stuff like upcast Hold Person didn't allow for target selection. So the full release included that new possibility, so they went ahead and changed a lot of the target spells to use that option.
Only downside is they went overboard with that and changed some spells that shouldn't include a target selection, such as Sleep, lol. I already reverted the Sleep changes on my mod, so no worries
